incrontab.1
.TH "incrontab" "1" "0.5.13" "Lukas Jelinek" "incron documentation"
.SH "NAME"
incrontab \- table manipulator for inotify cron (incron)
.SH "SYNOPSIS"
\fBincrontab\fR [\fB\-u\fR \fIuser\fR] [\fB\-f\fR \fIconfig\fR] \fIfile\fR

\fBincrontab\fR [\fB\-u\fR \fIuser\fR] [\fB\-f\fR \fIconfig\fR] [\fB\-l\fR | \fB\-r\fR | \fB\-e\fR | \fB\-t\fR | \fB\-d\fR]
.SH "DESCRIPTION"
incrontab is a table manipulator for the inotify cron (incron) system. It creates, removes, modifies and lists user tables (\fIincrontab\fR(5)).

Each user (including even system users without home directories) has an incron table which can't be manipulated directly (only root can effectively change these tables and is NOT recommended to do so).

All informational messages of this program are printed to the standard error output (stderr).

If \fI/etc/incron.allow\fR exists only users listed here may use incron. Otherwise if \fI/etc/incron.deny\fR exists only users NOT listed here may use incron. If none of these files exists everyone is allowed to use incron. (\fBImportant note:\fR This behavior is insecure and will be probably changed to be compatible with the style used by ISC Cron.) Location of these files can be changed in the configuration.

The first form of this command imports a file, validates it and stores to the table. "\-" can be used for loading from the standard input.



\fB\-u\fR (or \fB\-\-user\fR) option overrides the current (real) user to the given one. This option is intended for manipulation with system users' tables (such as apache, postfix, daemon etc.). It can be used only if the current user has root's effective rights.

\fB\-l\fR (or \fB\-\-list\fR) option causes the current table is printed to the standard output.

\fB\-r\fR (or \fB\-\-remove\fR) option causes the current table (if any) is permanently remove without any warning or confirmation. Use with caution!

\fB\-e\fR (or \fB\-\-edit\fR) option causes executing an editor for editing the user table (see below for the information about editor selection). You can edit your incron table now. If the table is changed it stores the modified version.

\fB\-t\fR (or \fB\-\-types\fR) option causes the list of supported event types (delimited by commas) is printed to the standard output. This feature is intended for front\-end applications to find out which event types was compiled in.

\fB\-d\fR (or \fB\-\-reload\fR) option causes reloading the current table by \fIincrond\fR(8). It is done through "touching" the table (writing into it without modifying it). This feature is intended e.g. for creating watches on newly created files (with already existing rules) or for rearming IN_ONESHOT watches.

\fB\-f <FILE>\fR (or \fB\-\-config=<FILE>\fR) option specifies another location for the configuration file (/etc/incron.conf is used by default). This feature requires root privileges.

There is a few complex algorithm how to determine which editor will be user for editing. If any of the following rule succeeds the appropriate editor is used:

1. EDITOR environment variable

2. VISUAL environment variable

3. configuration value

4. etc/alternatives/editor

5. hard\-wired editor (\fIvim\fR by default)

It's not recommended to use graphical editors (such as gVim, KEdit etc.) due to possible problems with connecting to the X server.
.SH "SEE ALSO"
incrond(8), incrontab(5), incron.conf(5)
.SH "AUTHOR"
Andreas Altair Redmer <altair.ibn.la.ahad.sy@gmail.com> (please report bugs to https://github.com/ar-/incron/issues ).
Lukas Jelinek <lukas@aiken.cz> .
.SH "COPYING"
This program is free software. It can be used, redistributed and/or modified under the terms of the GNU General Public License, version 2.


CHANGELOG
0.5.13   2021-08-17
* fix zombies without blocking too much
* new upstream release
* various changes

0.5.12   2015-02-01
* added recursive directory watching
* added configuration parameters: recursive, dotdirs and loopable
* added upstart script
* extended manpages and readme

0.5.11   2015-01-26
* first release from GitHub
* swiched all readme and manpage files to new upstream
* many bugfixes of improtant bugs
* tab delimiteres in config files allowed
* ignore dotfiles in /etc/incron.d
* added proper redirection and piping abilities to commands
* commands are now executed in a shel environment

0.5.10   2012-04-05
* some includes fixed to build correctly with GCC 4.6 (and 4.4-4.5 - #0000385, #0000447)
* some type issues fixed (#0000447)
* EAGAIN handling in poll() (#0000399, #0000686, #0000691)
* Makefile improved as of #0000512
* typo in various files fixed (#0000412)


0.5.9    2009-06-21
* spaces in event-related file names now handled correctly (#0000174)
* crashing on 'incrontab --user' fixed (#0000187)
* man pages improved (bug reporting info as of #0000407)
* Makefile improved ('release-clean' also removes 'doc')


0.5.8    2008-12-29
* more includes added for GCC 4.3 compatibility (#0000178)
* typo in help messages fixed (#0000172)
* some warnings fixed (#0000372)


0.5.7    2007-09-04
* man pages updated (mainly due to #0000157)
* some includes added for GCC 4.3 compatibility (#0000158)
* supplementary groups initialization (#0000167)


0.5.6    2007-04-18
* based on inotify-cxx 0.7.2 (fixes compilation problems on Debian Sarge, #0000155)


0.5.5    2007-02-14
* another bug related to environment variables fixed (#0000151)
* -V (--version) malfunction fixed (#0000150)


0.5.4    2007-02-11
* bug related to persisting environment fixed (#0000149)
* incrond(8) man page updated to explain how the environment is treated
* buggy example configuration file fixed


0.5.3    2007-02-08
* bug related to ignoring new tables fixed (#0000142)
* explicit typecasting in execlp() call (#0000145)
* memory leak related to loaded tables fixed (#0000144)
* documentation for 'incrontab -d' added (#0000146)
* man page for incron.conf added (#0000147)


0.5.2    2007-02-04
* lockfile permissions changed to 0644 (reopened #0000137)
* incorrect reporting of removing non-existing tables fixed (#0000141)


0.5.1    2007-01-28
* lockfile location problems fixed (#0000137)
* lockfile configuration failure fixed (#0000140)
* CRLFs inside LICENCE-GPL changed to LFs (#0000138)


0.5.0    2007-01-27
* based on inotify-cxx 0.7.1
* bug related to events names fixed (#0000134)
* instance locking (only one instance allowed - also fixes #0000136)
* configuration introduced (including custom cfg. files)
* waiting for child processes rewritten to avoid race conditions
* user can request reloading the table (#0000130)
* new implementation for command line arguments
* some code refactoring


0.4.0    2007-01-13
* based on inotify-cxx 0.7.0
* incrontab has a feature to find out supported event types (--types or -t)
* incrond now supports command line arguments
* incrond can be run on foreground (--foreground, -n)
* incrond can be simply stopped (--kill, -k)
* support for system tables (/etc/incron.d) added
* eaccess() replaced by euiaccess() (#0000125)
* each table now use its own inotify descriptor (#0000129)
* descriptors are closed on executing child processes


0.3.5    2007-01-09
* based on inotify-cxx 0.6.3
* a security bug related to access rights to watched files fixed (#0000119)
* crashing on table reload after removing a watched path fixed (#0000123)
* 'incrontab --list' dysfunction fixed (#0000122)


0.3.4    2007-01-03
* based on inotify-cxx 0.6.2
* problems with wrong ordered setting of UID/GID fixed (#0000115, #0000117)
* user tables have 0600 permission now


0.3.3    2006-12-15
* based on inotify-cxx 0.5.3
* "GID ignorance" bug fixed (#0000109)
* "zombie bug" fixed (#0000105)
* code cleanup and simplifying


0.3.2    2006-12-02
* problems with reiserfs fixed (#0000104)


0.3.1    2006-11-24
* user manual added
* incron tables now owned by root (also fixes #0000103)


0.3.0    2006-11-12
* based on inotify-cxx 0.5.2 (fixes a problem with ignoring IN_OPEN)
* two new special symbols ($% for flags as names, $& for numeric flags)
* whitespaces in paths now allowed (by prepending a backslash) (#0000098)
* some changes in logging etc.


0.2.3    2006-10-30
* problems with unwanted IN_NO_LOOP have been fixed (#0000097)


0.2.2    2006-10-29
* based on inotify-cxx 0.5.0
* 'uncleaned children' bug has been fixed (#0000095)
* memory leak (child process arguments) has been fixed (#0000096)


0.2.1    2006-10-14
* based on inotify-cxx 0.4.1 (it should also fix crashing as described in #0000085)
* build failure (due to missing 'useradd -M') on some distributions has been fixed
  (#0000094)
* tarball structure has been fixed (#0000089)
* documentation slightly improved


0.2.0    2006-10-04
* based on inotify-cxx 0.3.1
* better signal handling (no "dirty" wait() in a handler)
* loop avoidance can be used
* more documentation in the code


0.1.0    2006-09-15
* first alpha version


incrontab.cpp

/// inotify cron table manipulator classes implementation
/**
 * \file incrontab.cpp
 * 
 * inotify cron system
 * 
 * Copyright (C) 2006, 2007, 2008, 2012 Lukas Jelinek, <lukas@aiken.cz>
 * Copyright (C) 2014, 2015 Andreas Altair Redmer, <altair.ibn.la.ahad.sy@gmail.com>
 * 
 * This program is free software; you can use it, redistribute
 * it and/or modify it under the terms of the GNU General Public
 * License, version 2 (see LICENSE-GPL).
 *  
 * Credits:
 *   Christian Ruppert (new include to build with GCC 4.4+)
 * 
 */


#include <sstream>
#include <cstdio>
#include <errno.h>
//#include <syslog.h> // TODO remove

#include "inotify-cxx.h"

#include "incrontab.h"
#include "incroncfg.h"

#define IN_NO_LOOP_OLD "IN_NO_LOOP" // depricated from original version: no loop is now default
#define CT_LOOPABLE "loopable=true" // no loop is default. loopable must be set
#define CT_NORECURSION "recursive=false" // recursive is default, no recusion must be set
#define CT_DOTDIRS "dotdirs=true" // exclude dotdirs is default, include dotdirs must be set


/*
 * ALLOW/DENY SEMANTICS
 * 
 * If /etc/incron.allow exists ONLY users contained here
 * are allowed to use incron.
 * 
 * Otherwise, if /etc/incron.deny exists only user NOT
 * contained here are allowed to use incron.
 * 
 * Otherwise all users may use incron.
 * 
 */



IncronTabEntry::IncronTabEntry()
: m_uMask(0),
  m_fNoLoop(true),
  m_fNoRecursion(false),
  m_fDotDirs(false)
{
  
}

IncronTabEntry::IncronTabEntry(const std::string& rPath, uint32_t uMask, const std::string& rCmd)
: m_path(rPath),
  m_uMask(uMask),
  m_cmd(rCmd)
{
  
}

std::string IncronTabEntry::ToString() const
{
  std::ostringstream ss;
  
  std::string m;
  
  InotifyEvent::DumpTypes(m_uMask, m);
  // don't write IN_NO_LOOP anymore
//  if (m.empty())
//    m = m_fNoLoop ? IN_NO_LOOP_OLD : m;
//  else
//    if (m_fNoLoop) m.append(std::string(","+IN_NO_LOOP_OLD);

  // add CT_NORECURSION artificially
  if (m.empty())
    m = m_fNoRecursion ? CT_NORECURSION : m;
  else
    if (m_fNoRecursion) m.append(std::string(",")+CT_NORECURSION);
  
  // add CT_LOOPABLE artificially
  if (m.empty())
    m = !m_fNoLoop ? CT_LOOPABLE : m;
  else
    if (!m_fNoLoop) m.append(std::string(",")+CT_LOOPABLE);
  
  // add CT_DOTDIRS artificially
  if (m.empty())
    m = m_fDotDirs ? CT_DOTDIRS : m;
  else
    if (m_fDotDirs) m.append(std::string(",")+CT_DOTDIRS);
  
  // fill a default value for broken lines
  if (m.empty())
    m = "IN_ALL_EVENTS";
  
  ss << GetSafePath(m_path) << "\t" << m << "\t" << m_cmd;
  return ss.str();
}

bool IncronTabEntry::Parse(const std::string& rStr, IncronTabEntry& rEntry)
{
  unsigned long u;
  std::string s1, s2, s3;

  // ignore comment lines
  if (rStr.at(0) == '#')
    return false;

  StringTokenizer tok(rStr, " \t", '\\');
  if (!tok.HasMoreTokens())
    return false;
    
  s1 = tok.GetNextToken(true);
  if (!tok.HasMoreTokens())
    return false;
    
  s2 = tok.GetNextToken(true);
  if (!tok.HasMoreTokens())
    return false;
  
  tok.SetNoPrefix();
  s3 = tok.GetRemainder();
  SIZE len = s3.length();
  if (len > 0 && s3[len-1] == '\n')
    s3.resize(len-1);
  
  rEntry.m_path = s1;
  rEntry.m_cmd = s3;
  rEntry.m_uMask = 0;
  rEntry.m_fNoLoop = true;
  rEntry.m_fNoRecursion = false;
  rEntry.m_fDotDirs = false;
  
  if (sscanf(s2.c_str(), "%lu", &u) == 1) {
    rEntry.m_uMask = (uint32_t) u;
  }
  else {
    StringTokenizer tok2(s2);
    while (tok2.HasMoreTokens()) {
      std::string s(tok2.GetNextToken());
      if (s == IN_NO_LOOP_OLD)
        rEntry.m_fNoLoop = true;
      else if (s == CT_LOOPABLE)
        rEntry.m_fNoLoop = false;
      else if (s == CT_NORECURSION)
        rEntry.m_fNoRecursion = true;
      else if (s == CT_DOTDIRS)
        rEntry.m_fDotDirs = true;
      else
        rEntry.m_uMask |= InotifyEvent::GetMaskByName(s);
    }
  }
  
  return true;
}

std::string IncronTabEntry::GetSafePath(const std::string& rPath)
{
  std::ostringstream stream;
  
  SIZE len = rPath.length();
  for (SIZE i = 0; i < len; i++) {
    if (rPath[i] == ' ' ||
        rPath[i] == '\'' ||
        rPath[i] == '"' ||
        rPath[i] == '(' ||
        rPath[i] == ')' ||
        rPath[i] == '<' ||
        rPath[i] == '>' ||
        rPath[i] == '\\') {
      stream << "\\";
    }
    stream << rPath[i];
  }
  
  return stream.str();
}

bool IncronTab::Load(const std::string& rPath)
{
  m_tab.clear();
  
  FILE* f = fopen(rPath.c_str(), "r");
  if (f == NULL)
    return false;
  
  char s[1000];
  IncronTabEntry e;
  while (fgets(s, 1000, f) != NULL) {
    if (IncronTabEntry::Parse(s, e)) {
      m_tab.push_back(e);
    }
  }
  
  fclose(f);
  
  return true;
}

bool IncronTab::Save(const std::string& rPath)
{
  FILE* f = fopen(rPath.c_str(), "w");
  if (f == NULL)
    return false;
  
  std::deque<IncronTabEntry>::iterator it = m_tab.begin();
  while (it != m_tab.end()) {
    fputs((*it).ToString().c_str(), f);
    fputs("\n", f);
    it++;
  }
  
  fclose(f);
  
  return true;
}

bool IncronTab::CheckUser(const std::string& rUser)
{
  char s[100], u[100];
  
  std::string path;
  if (!IncronCfg::GetValue("allowed_users", path))
    throw InotifyException("configuration is corrupted", EINVAL);
  
  FILE* f = fopen(path.c_str(), "r");
  if (f == NULL) {
    if (errno == ENOENT) {
      if (!IncronCfg::GetValue("denied_users", path))
        throw InotifyException("configuration is corrupted", EINVAL);
      
      f = fopen(path.c_str(), "r");
      if (f == NULL) {
        return errno == ENOENT;
      }
      while (fgets(s, 100, f) != NULL) {
        if (sscanf(s, "%s", u) == 1) {
          if (rUser == u) {
            fclose(f);
            return false;
          }
        }
      }
      fclose(f);
      return true;
    }
    
    return false;
  }

  while (fgets(s, 100, f) != NULL) {
    if (sscanf(s, "%s", u) == 1) {
      if (rUser == u) {
        fclose(f);
        return true;
      }
    }
  }
  
  fclose(f);
  return false;
}

std::string IncronTab::GetUserTablePath(const std::string& rUser)
{
  std::string s;
  if (!IncronCfg::GetValue("user_table_dir", s))
    throw InotifyException("configuration is corrupted", EINVAL);
    
  return IncronCfg::BuildPath(s, rUser);
}

std::string IncronTab::GetSystemTablePath(const std::string& rName)
{
  std::string s;
  if (!IncronCfg::GetValue("system_table_dir", s))
    throw InotifyException("configuration is corrupted", EINVAL);
    
  return IncronCfg::BuildPath(s, rName);
}



